enterprise/backend/src/metabase_enterprise/gsheets/api.clj

```diff
@@ -204,116 +190,130 @@
+  (let [saved-setting (gsheets-safe)
+        conn-id (:gdrive/conn-id saved-setting)
+        hm-response (if (empty? saved-setting)
+                      nil
+                      (try (hm-get-gdrive-conn conn-id)
+                           (catch Exception _
+                             (reset-gsheets-status!)
+                             (error-response-in-body
+                              (tru "Unable to find google drive connection, please try again.")
+                              {:gdrive/conn-id conn-id}))))
+        [hm-status {hm-body :body}] hm-response]
+    (if (= :ok hm-status)
+      (let [{:keys [status last-sync-at last-sync-started-at]
+             :as   _} (normalize-gdrive-conn hm-body)]
         (-> (cond
-              (sync-complete? {:status status :last-dwh-sync last-dwh-sync :last-gdrive-conn-sync last-gdrive-conn-sync})
-              (u/prog1 (assoc (gsheets) :status "complete")
-                (gsheets! <>)
-                (analytics/inc! :metabase-gsheets/connection-creation-ok)
-                (snowplow/track-event! :snowplow/simple_event
-                                       {:event "sheets_connected" :event_detail "success"}))
-
-              ;; Timeout check
-              (when-let [upload-time (:folder-upload-time (gsheets))]
-                (> (seconds-from-epoch-now) (+ upload-time *folder-setup-timeout-seconds*)))
-              (do (reset-gsheets-status!)
-                  (analytics/inc! :metabase-gsheets/connection-creation-error {:reason "timeout"})
-                  (error-response-in-body (tru "Timeout syncing google drive folder, please try again.")
-                                          {:status-code 408}))
-
-              ;; Syncing failed
               (= "error" status)
               (do
                 (reset-gsheets-status!)
                 (analytics/inc! :metabase-gsheets/connection-creation-error {:reason "status_error"})
-                (error-response-in-body (tru "Problem syncing google drive folder, please try again..")))
-
-              ;; Continue waiting
-              :else (gsheets))
-            (assoc :db_id (:id attached-dwh)
-                   ;; here is some debugging info that proved very helpful:
-                   #_#_:hm/conn gdrive-conn
-                   #_#_:mb/sync-info {:status                status
-                                      :last-dwh-sync         last-dwh-sync
-                                      :last-gdrive-conn-sync last-gdrive-conn-sync})))
-      (do
-        (reset-gsheets-status!)
-        (error-response-in-body
-         (tru "Unable to find google drive connection.")
-         {:status-code 404})))))
-
-(api.macros/defendpoint :get "/folder" :- :gsheets/gsheets
+                (error-response-in-body (tru "Problem syncing google drive folder, please recreate the connection.")))
+
+              (= "active" status)
+              (assoc (setting->response saved-setting)
```

Comment:
I think you could move the `(assoc :db_id (:id attached-dwh))` from line 265 into the setting->response function, since it'll always be there.